drivers/usb/usbuhci_new/usbuhci.c

```diff
@@ -0,0 +1,2897 @@
+
+            EndpointRequirements->MaxTransferSize = UHCI_MAX_ISO_TRANSFER_SIZE;
+            break;
+
+        case USBPORT_TRANSFER_TYPE_CONTROL:
+            DPRINT("UhciQueryEndpointRequirements: ControlTransfer\n");
+            TdCont = EndpointProperties->MaxTransferSize /
+                     EndpointProperties->TotalMaxPacketSize;
+            TdCont += 2; // First + Last TDs
+
+            EndpointRequirements->HeaderBufferSize = sizeof(UHCI_HCD_QH) +
+                                      TdCont * sizeof(UHCI_HCD_TD);
+
+            EndpointRequirements->MaxTransferSize = EndpointProperties->MaxTransferSize;
+            break;
+
+        case USBPORT_TRANSFER_TYPE_BULK:
+            DPRINT("UhciQueryEndpointRequirements: BulkTransfer\n");
+            TdCont = 2 * UHCI_MAX_BULK_TRANSFER_SIZE /
+                     EndpointProperties->TotalMaxPacketSize;
+
+            EndpointRequirements->HeaderBufferSize = sizeof(UHCI_HCD_QH) +
+                                      TdCont * sizeof(UHCI_HCD_TD);
+
+            EndpointRequirements->MaxTransferSize = UHCI_MAX_BULK_TRANSFER_SIZE;
+            break;
+
+        case USBPORT_TRANSFER_TYPE_INTERRUPT:
+            DPRINT("UhciQueryEndpointRequirements: InterruptTransfer\n");
+            TdCont = 2 * UHCI_MAX_INTERRUPT_TD_COUNT;
+
+            EndpointRequirements->HeaderBufferSize = sizeof(UHCI_HCD_QH) +
+                                      TdCont * sizeof(UHCI_HCD_TD);
+
+            EndpointRequirements->MaxTransferSize = UHCI_MAX_INTERRUPT_TD_COUNT *
+                                      EndpointProperties->TotalMaxPacketSize;
+            break;
+
+        default:
+            DPRINT1("UhciQueryEndpointRequirements: Unknown TransferType - %x\n",
+                    TransferType);
+            DbgBreakPoint();
+            break;
+    }
+}
+
+VOID
+NTAPI
+UhciCloseEndpoint(IN PVOID uhciExtension,
+                  IN PVOID uhciEndpoint,
+                  IN BOOLEAN IsDoDisablePeriodic)
+{
+    DPRINT_IMPL("UhciCloseEndpoint: UNIMPLEMENTED. FIXME\n");
+}
+
+MPSTATUS
+NTAPI
+UhciTakeControlHC(IN PUHCI_EXTENSION UhciExtension,
+                  IN PUSBPORT_RESOURCES Resources)
+{
+    LARGE_INTEGER EndTime;
+    LARGE_INTEGER CurrentTime;
+    ULONG ResourcesTypes;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT StatusRegister;
+    UHCI_PCI_LEGSUP LegacySupport;
+    UHCI_PCI_LEGSUP LegacyMask;
+    UHCI_USB_COMMAND Command;
+    UHCI_USB_STATUS HcStatus;
+    MPSTATUS MpStatus = MP_STATUS_SUCCESS;
+
+    DPRINT("UhciTakeControlHC: Resources->ResourcesTypes - %x\n",
+           Resources->ResourcesTypes);
+
+    ResourcesTypes = Resources->ResourcesTypes;
+
+    if ((ResourcesTypes & (USBPORT_RESOURCES_PORT | USBPORT_RESOURCES_INTERRUPT)) !=
+                          (USBPORT_RESOURCES_PORT | USBPORT_RESOURCES_INTERRUPT))
+    {
+        DPRINT1("UhciTakeControlHC: MP_STATUS_ERROR\n");
+        MpStatus = MP_STATUS_ERROR;
+    }
```

Comment:
Should the function return here? Is there other error handling missing?

Reply:
No. Here it should be. It ok.